Add scripts to manage SV spark jobs and copy result

TedBrookings opened the pull request:
Add new scripts to gatk/scripts/sv/ folder, and alter action (but not
passed parameters) of older scripts to make running sv spark jobs
more convenient.
Added:
  -copy_sv_results.sh: copy files to time and git-stamped folder on
   google cloud storage
     -> results folder on cluster
     -> command line arguments to SV discover pipeline
     -> console log file (if present)

  -manage_sv_pipeline.sh: create cluster, run job, copy results, and
   delete cluster. Manage cluster naming, time and git-stamping,
   and log file production.

Altered:
  -create_cluster.sh: control GCS zone and numbers of workers via
   environmental variables. Defaults to previous hard-coded values.

  -runWholePipeline.sh: accept command-line arguments for sv
   discovery pipeline, work with clusters having NUM_WORKERS != 10

SHuang-Broad commented on scripts/sv/runWholePipeline.sh:
```diff
@@ -34,6 +35,16 @@ KMER_KILL_LIST=$(echo "${REF_FASTA}" | sed 's/.fasta$/.kill.kmers/')
+eval "SV_ARGS=\"${SV_ARGS}\""
+
+# Choose NUM_EXECUTORS = 2 * NUM_WORKERS
+NUM_WORKERS=$(gcloud compute instances list --filter="name ~ ${CLUSTER_NAME}-[sw].*" | grep RUNNING | wc -l)
```
I got warnings saying "This argument is deprecated". 
How about
```bash
gcloud dataproc clusters list | grep -F "${CLUSTER_NAME}" | awk '{print $2}'
```

TedBrookings replied:
That would eliminate the warning.  I think the warning itself is actually a bug (it corresponds to passing the NAME argument, not using a filter). I've avoided listing all instances and using grep because that corresponds to potentially downloading a large number of instances, whereas this in principle might filter before download, saving data. If the warning is disconcerting, I can make the fix you suggest and then revert once google gets their act together on this.

TedBrookings replied:
I found a filter command for the gcloud dataproc clusters list command, so I changed and pushed that. It has the downside of not seeing preemptible instances, but preemptible instances actually crash the pipeline right now, so it's not a super-important problem.

SHuang-Broad commented on scripts/sv/manage_sv_pipeline.sh:
```diff
@@ -0,0 +1,174 @@
+  [4] GCS path to reference fasta (required)
+      OPTIONAL arguments:
+  [5] path to initialization script (local or GCS, defaults to
+      ${GATK_DIR}/scripts/sv/default_init.sh if omitted or empty)
```
I have a concern here: if the 5th and 6th argument are omitted, is the `shift 6` statement below going to skip two arguments that are intended to be sent to java be "shifted" away?

TedBrookings replied:
The user would have to pass empty strings (signaling default values) for argument 4 and 5 to then pass arguments 6 and above. I added a description of this to the documentation and pushed the fix.

SHuang-Broad commented on scripts/sv/manage_sv_pipeline.sh:
```diff
@@ -0,0 +1,174 @@
+PATH="${GATK_DIR}/scripts/sv:${PATH}"
+
+# configure caching .jar files
+export GATK_GCS_STAGING="gs://${PROJECT_NAME}/${GCS_USER}/staging/"
```
I would prefer a check if an environment variable has been set or not, because a user may already have this set in my shell profile ( I personally do), and preferably would like the jars being cached there.

TedBrookings replied:
Agreed. I'll fix that.

SHuang-Broad commented on scripts/sv/manage_sv_pipeline.sh:
```diff
@@ -0,0 +1,174 @@
+export GATK_GCS_STAGING="gs://${PROJECT_NAME}/${GCS_USER}/staging/"
+
+# set cluster name based on user and target bam file
+# (NOTE: can override by defining SV_CLUSTER_NAME)
```
This doesn't seem to be exposed to the user (i.e. it is hidden knowledge).

TedBrookings replied:
Sorry, I somehow missed this comment. I just added a line to inform the user what the CLUSTER_NAME was set to. I'll push again.

SHuang-Broad commented on scripts/sv/manage_sv_pipeline.sh:
```diff
@@ -0,0 +1,174 @@
+CLUSTER_NAME=${SV_CLUSTER_NAME:-"${GCS_USER}-${SANITIZED_BAM}"}
+
+# update gcloud
+gcloud components update
```
requires interaction from user, may be a surprise to someone.

SHuang-Broad commented on scripts/sv/manage_sv_pipeline.sh:
```diff
@@ -0,0 +1,174 @@
+# copy results into gcloud
+while true; do
+    echo "#############################################################" 2>&1 | tee -a ${LOCAL_LOG_FILE}
+    read -p "Copy results? (yes/no/cancel)" yn
```
sorry for being edge-case picky: if a user answered no to run whole pipeline, this would end up copying nothing, correct?

TedBrookings replied:
It depends. They may have run previously (perhaps not using this script, or perhaps there was an error). If there's data, the copy script should copy it off, and regardless, it should copy any log files or command line arguments.

SHuang-Broad commented on scripts/sv/copy_sv_results.sh:
```diff
@@ -0,0 +1,98 @@
+GCS_USER=${4:-${USER}}
+LOCAL_LOG_FILE=${5:-"/dev/null"}
+
+shift 5
```
Same concern: if we only provide 3 required arguments and the rest is java arguments, are 4 and 5 "shifted" away.

TedBrookings replied:
Here as well, I just added a description of how to specify later values and leave early ones as default.

SHuang-Broad commented:
Turns out a typo prevents running the "manage_sv_pipeline" script, saying GATK_DIR is an unbound variable. Please fix.

TedBrookings commented:
Unbound variable bug fixed.

SHuang-Broad commented:
I cannot seem to be able to run the pipeline with the following way of running it:

```bash
./manage_sv_pipeline.sh \
    /Users/shuang/GATK \
    broad-dsde-methods \
    gs://broad-dsde-methods/sv/samples/G94797_CHM_MIX/WGS1/G94794.CHMI_CHMI3_WGS1.cram.bam \
    gs://broad-dsde-methods/sv/reference/GRCh38/Homo_sapiens_assembly38.fasta \
    gs://broad-dsde-methods/shuang/tmp/gatk-jars/default_init.sh
```

And if I run with the following way (only adding a user name)

```bash
./manage_sv_pipeline.sh \
    /Users/shuang/GATK \
    broad-dsde-methods \
    gs://broad-dsde-methods/sv/samples/G94797_CHM_MIX/WGS1/G94794.CHMI_CHMI3_WGS1.cram.bam \
    gs://broad-dsde-methods/sv/reference/GRCh38/Homo_sapiens_assembly38.fasta \
    gs://broad-dsde-methods/shuang/tmp/gatk-jars/default_init.sh \
    shuang
```
The script runs as expected.

I believe line 47 is the culprit:
```bash
SV_ARGS=${*:-${SV_ARGS:-""}} && SV_ARGS=${SV_ARGS:+" ${SV_ARGS}"}
```

SHuang-Broad commented:
My test run has successfully finished. Awesome, @TedBrookings !

Two suggestions that you could decide when to address them:
1. This script requires a lot of user interaction (gcloud update, whether to create a cluster, whether to copy results, whether to delete cluster), I could imagine this being inconvenient for some, so we could have some upfront arguments specifying answers to these questions when the script is launched.
2. The bucket to which the results are copied is not overridable by caller of script. A user might want to copy the results to a specific location.

TedBrookings commented:
I addressed the crash (caused by calling shift with too large a number and set-ing strict error checks in bash).

I also added the requested options. Due to the larger number of options I decided that using only positional arguments would be impractical (requiring counting the number of empty "" default arguments) so I added a unix-like option parsing scheme to manage_sv_pipeline.sh
--quiet or -q will cause the script to run without any user prompting
--save [bucket/path] or -s [bucket/path] will save results and initialization scripts within the specified bucket/path
